README.md
# PHP IP Library

IPv4/IPv6 manipulation library for PHP inspired by Python [ipaddress](https://docs.python.org/dev/library/ipaddress.html).

[![Build status](https://github.com/rlanvin/php-ip/workflows/Tests/badge.svg)](https://github.com/rlanvin/php-ip/actions)
[![Latest Stable Version](https://poser.pugx.org/rlanvin/php-ip/v/stable)](https://packagist.org/packages/rlanvin/php-ip)
[![Total Downloads](https://poser.pugx.org/rlanvin/php-ip/downloads)](https://packagist.org/packages/rlanvin/php-ip)

## Requirements

- PHP >= 7.1
- IPv6 support enabled
- GMP extension (www.php.net/manual/en/book.gmp.php)

## Installation

The recommended way is to install the lib [through Composer](http://getcomposer.org/).

Simply run `composer require rlanvin/php-ip` for it to be automatically installed and included in your `composer.json`.

Now you can use the autoloader, and you will have access to the library:

```php
require 'vendor/autoload.php';
```

## Documentation

Complete doc is available in [the wiki](https://github.com/rlanvin/php-ip/wiki).

## Related projects

There are some Open Source libraries, built by community on top of this this one:

- [hiqdev/php-ip-tools](https://github.com/hiqdev/php-ip-tools) – tooling for IP address calculations, such as free IP blocks calculation and IP ranges parsing.
- Built one? Add it here.

## Contribution

Feel free to contribute! Just create a new issue or a new pull request.

## License

This library is released under the MIT License.


CHANGELOG.md
# Changelog

## [Unreleased]

## [3.0.0] - 2022-03-02

### Backward Compatibility Breaking Changes

- Dropped support for PHP 7.0

### Added

- Support for PHP 8.1


## [3.0.0-rc2] - 2021-07-22

### Backward Compatibility Breaking Changes

- Separate "[private](https://en.wikipedia.org/wiki/Private_network)" from other "[reserved](https://en.wikipedia.org/wiki/Reserved_IP_addresses)" IP addresses.
This changes the behaviour of `IP::isPrivate()` and `IPBlock::getPrivateBlocks()` to be more narrow in scope.
The previous behaviour has been moved to new methods: `IP::isReserved()` and `IPBlock::getReservedBlocks()`.
- `IP::isPrivate()` renamed to `IP::isReserved()`
- `IPBlock::getPrivateBlocks()` renamed to `IPBlock::getReservedBlocks()`
- `IP::isPrivate()` now only tests if an IP is in a forwardable and non globally reachable IP Block as defined in the IANA special-purpose address registry
- `IPBlock::getPrivateBlocks()` now returns only the forwardable and non globally reachable IP Blocks as defined in the IANA special-purpose address registry


## [3.0.0-rc1] - 2021-04-18

### Backward Compatibility Breaking Changes

- Deprecated auto-detection of binary strings for IPv6 [#57](https://github.com/rlanvin/php-ip/pull/57)
This means you can't directly pass the result of `inet_pton` to construct an IPv6 instance anymore, because it is
impossible to reliability distinguish some IPv6 human-readable representation from their binary string representation.
Instead, you need to use the explicit factory method `IPv6::createFromBinaryString` if you want to work with `inet_pton`.
Note: this is still supported for IPv4.
- `IPBlock::getMask()` renamed to `IPBlock::getNetmask()`
- `IPBlock::getPrefix()` renamed to `IPBlock::getPrefixLength()`
- `IPBlock::getMaxPrefix()` renamed to `IPBlock::getMaxPrefixLength()`
- `IPBlock::getGivenIpWithPrefixLen()` renamed to `IPBlock::getGivenIpWithPrefixLength()`

### Added

- Support for PHP 8.0
- Added explicit factory methods for `IPv4`/`IPv6` classes:
`createFromInt`, `createFromFloat`, `createFromString`, `createFromBinaryString`, `createFromNumericString` and `createFromGmp`
- `IPBlock::plus()` and `IpBlock::minus()` now accept numeric strings and GMP instances
- `IPBlockIterator` now implements `ArrayAccess` which means you can write e.g. `IPBlock::create('192.168.0.0/24')->getSubBlocks('/25')[1]`
- New methods for `IPv4Block`/`IPv6Block`: `getPrivateBlocks()`, `getLoopbackBlock()`, `getLinkLocalBlock()`

### Changed

- Optimise performance of `IP::isPrivate()`

### Fixed

- `IPBlock::contains` now always throws an `InvalidArgumentException` if mixing IP versions

### Removed

- Removed deprecated method `IPBlock::getSuper`

## [2.1.0] - 2020-10-31

### Added

- New method `IP::matches($ip, $mask)` to perform wildcard mask matching common in network Access Control Lists and OSPF dynamic routing [#51](https://github.com/rlanvin/php-ip/pull/51)
- IPv4Block: Allow to specify the prefix also as an old-style netmask [#53](https://github.com/rlanvin/php-ip/pull/53)
- Support for PHP 7.4

## [2.0.0] - 2019-09-01

### Backward Compatibility Breaking Changes

- All classes now reside within `PhpIP` namespace. [#14](https://github.com/rlanvin/php-ip/pull/14)
- Deprecated `IPBlock::getSuper`, use `IPBlock::getSuperBlock()` instead. [#46](https://github.com/rlanvin/php-ip/pull/46)

### Added

- New method `isLoopback()` [#37](https://github.com/rlanvin/php-ip/pull/37)
- New method `isLinkLocal()` [#43](https://github.com/rlanvin/php-ip/pull/43)
- New method `reversePointer()` [#44](https://github.com/rlanvin/php-ip/pull/44)
- New method `IpBlock::getGivenIp` to return the IP used in the constructor [#6](https://github.com/rlanvin/php-ip/pull/6)

### Removed

- Drop support for PHP 5 [#8](https://github.com/rlanvin/php-ip/issues/8)

## [1.0.1] - 2015-06-26

### Fixed

- Compatibility issues with PHP 5.4, 5.5 and 5.6

## 1.0.0 - 2015-04-03

First release

[Unreleased]: https://github.com/rlanvin/php-ip/compare/v3.0.0...HEAD
[3.0.0]: https://github.com/rlanvin/php-ip/compare/v3.0.0-rc2...v3.0.0
[3.0.0-rc2]: https://github.com/rlanvin/php-ip/compare/v3.0.0-rc1...v3.0.0-rc2
[3.0.0-rc1]: https://github.com/rlanvin/php-ip/compare/v2.1.0...v3.0.0-rc1
[2.1.0]: https://github.com/rlanvin/php-ip/compare/v2.0.0...v2.1.0
[2.0.0]: https://github.com/rlanvin/php-ip/compare/v1.0.1...v2.0.0
[1.0.1]: https://github.com/rlanvin/php-ip/compare/v1.0.0...v1.0.1


src/IPv4Block.php
<?php

declare(strict_types=1);

/**
 * Licensed under the MIT license.
 *
 * For the full copyright and license information, please view the LICENSE file.
 *
 * @author Rémi Lanvin <remi@cloudconnected.fr>
 *
 * @see https://github.com/rlanvin/php-ip
 */

namespace PhpIP;

/**
 * An IPv4 CIDR block.
 */
class IPv4Block extends IPBlock
{
    use IPBlockTrait;

    const IP_CLASS = IPv4::class;

    /**
     * @see https://www.iana.org/assignments/iana-ipv4-special-registry/iana-ipv4-special-registry.xhtml
     *
     * PRIVATE_BLOCKS   = not globally reachable, but routeable
     * RESERVED_BLOCKS  = all IANA reserved blocks
     * LOOPBACK_BLOCK   = Loopback Address
     * LINK_LOCAL_BLOCK = Link-Local Unicast
     */
    const PRIVATE_BLOCKS = [
        '10.0.0.0/8', // Private-Use
        '100.64.0.0/10', // Shared Address Space (for communications between a service provider and its subscribers when using a carrier-grade NAT)
        '172.16.0.0/12', // Private-Use
        '192.0.0.0/29', // IPv4 Service Continuity Prefix (Dual-Stack Lite)
        '192.168.0.0/16', // Private-Use
        '198.18.0.0/15' // Benchmarking (testing of inter-network communications between two separate subnets)
    ];
    const RESERVED_BLOCKS = [
        '0.0.0.0/8',
        '0.0.0.0/32',
        '10.0.0.0/8',
        '100.64.0.0/10',
        '127.0.0.0/8',
        '169.254.0.0/16',
        '172.16.0.0/12',
        '192.0.0.0/24',
        '192.0.0.0/29',
        '192.0.0.8/32',
        '192.0.0.9/32',
        '192.0.0.10/32',
        '192.0.0.170/32',
        '192.0.0.171/32',
        '192.0.2.0/24',
        '192.31.196.0/24',
        '192.52.193.0/24',
        '192.88.99.0/24',
        '192.168.0.0/16',
        '192.175.48.0/24',
        '198.18.0.0/15',
        '198.51.100.0/24',
        '203.0.113.0/24',
        '240.0.0.0/4',
        '255.255.255.255/32',
    ];
    const LOOPBACK_BLOCK = '127.0.0.0/8';
    const LINK_LOCAL_BLOCK = '169.254.0.0/16';

    const NETMASK2PREFIX = [
        '0.0.0.0' => 0,
        '128.0.0.0' => 1,
        '192.0.0.0' => 2,
        '224.0.0.0' => 3,
        '240.0.0.0' => 4,
        '248.0.0.0' => 5,
        '252.0.0.0' => 6,
        '254.0.0.0' => 7,
        '255.0.0.0' => 8,
        '255.128.0.0' => 9,
        '255.192.0.0' => 10,
        '255.224.0.0' => 11,
        '255.240.0.0' => 12,
        '255.248.0.0' => 13,
        '255.252.0.0' => 14,
        '255.254.0.0' => 15,
        '255.255.0.0' => 16,
        '255.255.128.0' => 17,
        '255.255.192.0' => 18,
        '255.255.224.0' => 19,
        '255.255.240.0' => 20,
        '255.255.248.0' => 21,
        '255.255.252.0' => 22,
        '255.255.254.0' => 23,
        '255.255.255.0' => 24,
        '255.255.255.128' => 25,
        '255.255.255.192' => 26,
        '255.255.255.224' => 27,
        '255.255.255.240' => 28,
        '255.255.255.248' => 29,
        '255.255.255.252' => 30,
        '255.255.255.254' => 31,
        '255.255.255.255' => 32,
    ];

    /**
     * @internal
     * Check if the prefix length is valid,
     * rewrites IPv4 old-style netmask to CIDR prefix length,
     * and returns the prefix length as an int
     *
     * @param mixed $prefix
     *
     * @throws \InvalidArgumentException
     *
     * @return int
     */
    protected function checkPrefixLength($prefix)
    {
        if (is_string($prefix) && isset(self::NETMASK2PREFIX[$prefix])) {
            return (int) self::NETMASK2PREFIX[$prefix];
        }

        return parent::checkPrefixLength($prefix);
    }
}
